src/Workspaces/Remote/ServiceHub/Host/SolutionAssetCache.cs

```diff
@@ -11,230 +11,263 @@
+            compilationStateChecksums.AddAllTo(pinnedChecksums);
 
-            using (Logger.LogBlock(FunctionId.AssetStorage_CleanAssets, cancellationToken))
+            var solutionState = compilationState.SolutionState;
+            Contract.ThrowIfFalse(solutionState.TryGetStateChecksums(out var stateChecksums));
+            stateChecksums.AddAllTo(pinnedChecksums);
+
+            foreach (var (_, projectState) in solutionState.ProjectStates)
             {
-                // Ensure that if our remote workspace has a current solution, that we don't purge any items associated
-                // with that solution.
-                PooledHashSet<Checksum>? pinnedChecksums = null;
-                try
-                {
-                    foreach (var (checksum, entry) in _assets)
-                    {
-                        // If not enough time has passed, keep in the cache.
-                        if (entry.Stopwatch.Elapsed <= _purgeAfterTimeSpan)
-                            continue;
-
-                        // If this is a checksum we want to pin, do not remove it.
-                        if (pinnedChecksums == null)
-                        {
-                            pinnedChecksums = PooledHashSet<Checksum>.GetInstance();
-                            await AddPinnedChecksumsAsync(pinnedChecksums, cancellationToken).ConfigureAwait(false);
-                        }
-
-                        if (pinnedChecksums.Contains(checksum))
-                            continue;
-
-                        _assets.TryRemove(checksum, out _);
-                    }
-                }
-                finally
-                {
-                    pinnedChecksums?.Free();
-                }
+                Contract.ThrowIfFalse(projectState.TryGetStateChecksums(out var projectStateChecksums));
+                projectStateChecksums.AddAllTo(pinnedChecksums);
+
+                AddAll(projectState.DocumentStates);
+                AddAll(projectState.AdditionalDocumentStates);
+                AddAll(projectState.AnalyzerConfigDocumentStates);
             }
         }
 
-        private async ValueTask AddPinnedChecksumsAsync(HashSet<Checksum> pinnedChecksums, CancellationToken cancellationToken)
+        void AddAll<TState>(TextDocumentStates<TState> states) where TState : TextDocumentState
         {
-            if (_remoteWorkspace is null)
-                return;
-
-            var checksums = await _remoteWorkspace.CurrentSolution.CompilationState.GetStateChecksumsAsync(cancellationToken).ConfigureAwait(false);
-            checksums.AddAllTo(pinnedChecksums);
+            foreach (var (_, documentState) in states.States)
+            {
+                Contract.ThrowIfFalse(documentState.TryGetStateChecksums(out var documentChecksums));
+                documentChecksums.AddAllTo(pinnedChecksums);
```

Comment:
importantly, this will prevent us from tossing away checksum->document mappings in the asset cache on one of the manual-GC passes it does, thus forcing it to have to be retrieved in the future.